packages/core/src/messages/responses/storage/scalar/storage-get.ts

```diff
@@ -1,142 +1,159 @@
+  // readonly itemType: StorageItemType;
 
-  abstract value(): string | number | Uint8Array;
-
-  abstract valueInt(): number | undefined;
-
-  abstract valueDouble(): number | undefined;
-
-  abstract valueString(): string | undefined;
-
-  abstract valueBytes(): Uint8Array | undefined;
-}
-
-export class StringResponse extends Success {
-  override readonly itemType: StorageItemType.String = StorageItemType.String;
-  private readonly _value: string;
-  constructor(value: string) {
+  constructor(value: StorageValue | undefined) {
     super();
     this._value = value;
   }
 
-  value(): string {
-    return this._value;
+  static ofInt(value: number): Success {
+    return new Success(StorageValue.ofInt(value));
   }
 
-  valueBytes(): undefined {
-    return undefined;
+  static ofDouble(value: number): Success {
+    return new Success(StorageValue.ofDouble(value));
   }
 
-  valueDouble(): undefined {
-    return undefined;
+  static ofString(value: string): Success {
+    return new Success(StorageValue.ofString(value));
   }
 
-  valueInt(): undefined {
-    return undefined;
-  }
-
-  valueString(): string {
-    return this.value();
-  }
-}
-
-export class IntegerResponse extends Success {
-  override readonly itemType: StorageItemType.Integer = StorageItemType.Integer;
-  private readonly _value: number;
-  constructor(value: number) {
-    super();
-    this._value = value;
+  static ofBytes(value: Uint8Array): Success {
+    return new Success(StorageValue.ofBytes(value));
   }
 
-  value(): number {
+  value(): StorageValue | undefined {
     return this._value;
   }
-
-  valueBytes(): undefined {
-    return undefined;
-  }
-
-  valueDouble(): undefined {
-    return undefined;
-  }
-
-  valueInt(): number | undefined {
-    return this.value();
-  }
-
-  valueString(): undefined {
-    return undefined;
-  }
 }
 
-export class DoubleResponse extends Success {
-  override readonly itemType: StorageItemType.Double = StorageItemType.Double;
-  private readonly _value: number;
-  constructor(value: number) {
-    super();
-    this._value = value;
-  }
-
-  value(): number {
-    return this._value;
-  }
-
-  valueBytes(): undefined {
-    return undefined;
-  }
-
-  valueDouble(): number {
-    return this.value();
-  }
-
-  valueInt(): undefined {
-    return undefined;
-  }
-
-  valueString(): undefined {
-    return undefined;
-  }
-}
-
-export class BytesResponse extends Success {
-  override readonly itemType: StorageItemType.Bytes = StorageItemType.Bytes;
-  private readonly _value: Uint8Array;
-  constructor(value: Uint8Array) {
-    super();
-    this._value = value;
-  }
-
-  value(): Uint8Array {
-    return this._value;
-  }
-
-  valueBytes(): Uint8Array {
-    return this.value();
-  }
-
-  valueDouble(): undefined {
-    return undefined;
-  }
-
-  valueInt(): undefined {
-    return undefined;
-  }
-
-  valueString(): undefined {
-    return undefined;
-  }
-}
+// export class StringResponse extends Success {
```

Comment:
pulling this stuff out into a separate StorageValue class makes the logic simpler, and makes it possible / much easier to distinguish between NOT FOUND and "you chose the wrong type"